nixos/lib/test-driver/lib/driver.py

```diff
@@ -0,0 +1,165 @@
+
+    def __init__(
+        self,
+        logger: Logger,
+        vm_scripts: List[str],
+        tests: str,
+        interactive: bool = False,
+        keep_vm_state: bool = False,
+        configure_python_repl: Callable = id,
+        machine_class: Type[Machine] = Machine,
+        vlan_class: Type[VLan] = VLan,
+    ):
+        """
+        Args:
+            - configure_python_repl: a function to configure ptpython.repl
+        """
+        self.log = logger
+        self.configure_python_repl = configure_python_repl
+        self.tests = tests
+        self.interactive = interactive
+
+        tmp_dir = Path(os.environ.get("TMPDIR", tempfile.gettempdir()))
+        tmp_dir.mkdir(mode=0o700, exist_ok=True)
+
+        self.vlans = [
+            vlan_class(int(nr), tmp_dir, logger.log_machinestate)
+            for nr in list(dict.fromkeys(os.environ.get("VLANS", "").split()))
+        ]
+
+        def cmd(scripts: List[str]) -> Iterator[NixStartScript]:
+            for s in scripts:
+                yield NixStartScript(s)
+
+        self.machines = [
+            machine_class(
+                start_command=cmd,
+                keep_vm_state=keep_vm_state,
+                name=cmd.machine_name,
+                log_serial=logger.log_serial,
+                log_machinestate=logger.log_machinestate,
+                tmp_dir=tmp_dir,
+            )
+            for cmd in cmd(vm_scripts)
+        ]
+
+        @atexit.register
+        def clean_up() -> None:
+            with self.log.nested("clean up"):
+                for machine in self.machines:
+                    machine.release()
+                for vlan in self.vlans:
+                    vlan.release()
+            self.log.release()
+
+    def subtest(self, name: str) -> Iterator[None]:
+        """Group logs under a given test name"""
+        with self.log.nested(name):
+            try:
+                yield
+                return True
+            except:
+                self.log(f'Test "{name}" failed with error:')
+                raise
+
+    def test_symbols(self) -> Dict[str, Any]:
+        @contextmanager
+        def subtest(name: str) -> Iterator[None]:
+            return self.subtest(name)
+
+        general_symbols = dict(
+            pprint=pprint,
+            os=os,
+            driver=self,
+            logger=self.log,
+            vlans=self.vlans,
+            machines=self.machines,
+            start_all=self.start_all,
+            test_script=self.test_script,
+            subtest=subtest,
+        )
+        machine_symbols = {
+            m.name: self.machines[idx] for idx, m in enumerate(self.machines)
+        }
+        vlan_symbols = {
+            f"vlan{v.nr}": self.vlans[idx] for idx, v in enumerate(self.vlans)
+        }
+        print(
+            "Available Symbols:\n    "
+            + ", ".join(map(lambda m: m.name, self.machines))
+            + "  -  "
+            + ", ".join(map(lambda v: f"vlan{v.nr}", self.vlans))
```

Comment:
unfortunately this makes not much sense to me. why would i want to manage the lans myself? i want them to be just wired so i can fire up any vm whenever i want. please just implement the vlan class in a way that it runs the process in the constructor and then we're done with vlans. vlans are cheap.

Reply:
I want to explore ways to play chaos monkey or something like that. This is a start.

I also want to explore ways to interact with `vdeterm` in the future. This is a start.

Conclusion: this is a start. :wink:

Reply:
I don't agree that this should take place in a pull request that ought to refactor the nixos integration test driver as prior work towards having nice tmux-vm-tabs.

if you want to introduce new fun concepts, then please do that in follow-up pull requests. i am totally open to new concepts, but what happens here in this PR is feature creep.

Reply:
Alright, then I'll rather give you a specific reason to your question:

> why would i want to manage the lans myself?

In this refactoring is included the goal of doing away with semantical imprecision and to improve the UX.

- `start_all` in it's previous form technically was a lie. For this refactoring it is irrelevant if people have become accustomed to that lie.
- there has never been any explicit (enough) notion of the vlans. If you where new to the testing infrastructure, it has been easy to overlook what's going on with regard to vlans.

Both of the above points are sufficiently addressed by this refactoring. Reverting it, would re-introduce those issues and invalidate the goals of this refactoring.

Reply:
I'm going to postpone this discussion, please feel free to retake it in the future. For now, I need to focus on the bigger picture and I have a sense, we need to ensure overall correct program execution.

Reply:
you did more than just refactor the code, and now you find reasons for it a posteriori. "The network lie" was part of the interface before and changing that in a refactoring doesn't make sense to me. There is also no reason to change this at all.

Reply:
> There is also no reason to change this at all.

Well, if one looks closely enough through our past discussion, one will be able to spot a few. One might agree/disagree that those are valid and need be acted upon.